File: features/miso/loop/cards/browse/map-only/since/since.js
// the day starts, computed where the clock and the zone are. A wasm build has
// neither SystemTime nor a local time zone (/browse's own date arithmetic says
// so), so it cannot know when local midnight was; the page does. The three
// marks travel in as ONE event — the way a finger sends one — rather than as a
// write to a bridged key, because a node newer than /payload moves state with
// events and never by writing a bridged key (misses.md, navigation from the
// wrong side).
const feature_Since = {
  sent: '',
  midnight: null,

  // midnight today; midnight of the most recent Monday (Monday is day 1, so a
  // Sunday goes back six days); midnight of the 1st. Built by walking a Date
  // rather than by subtracting milliseconds, so a clock change inside the week
  // or the month lands on the wall-clock midnight and not an hour either side.
  marks() {
    const day = new Date();
    day.setHours(0, 0, 0, 0);
    const week = new Date(day.getTime());
    week.setDate(week.getDate() - ((week.getDay() + 6) % 7));
    week.setHours(0, 0, 0, 0);
    const month = new Date(day.getTime());
    month.setDate(1);
    month.setHours(0, 0, 0, 0);
    return day.getTime() + ',' + week.getTime() + ',' + month.getTime();
  },

  // quiet when nothing moved: a visibility flip on a phone that has not
  // crossed midnight or a time zone costs no event and no repaint.
  tell() {
    if (typeof feature_Loop === 'undefined') return;
    const m = this.marks();
    if (m === this.sent) { this.arm(); return; }
    this.sent = m;
    try { feature_Loop.send({ type: 'SinceMarks', data: { marks: m } }); } catch (e) { /* not up yet */ }
    this.arm();
  },

  // one timer, aimed at the next local midnight plus a second. Re-aimed every
  // time it fires, so a phone left open overnight rolls over on its own.
  arm() {
    if (this.midnight) clearTimeout(this.midnight);
    const next = new Date();
    next.setHours(24, 0, 1, 0);
    let wait = next.getTime() - Date.now();
    if (!isFinite(wait) || wait < 1000) wait = 1000;
    if (wait > 86400000) wait = 86400000;
    this.midnight = setTimeout(() => this.tell(), wait);
  },

  // the words the long-press card says for the four pills. /tool-words keeps
  // the same table for the tools and the picker's buttons; these are this
  // node's own, so they arrive and leave with it.
  WORDS: {
    since_today: { name: 'today', intro: 'Only what happened since midnight.' },
    since_week: { name: 'week', intro: 'This week, from Monday morning.' },
    since_month: { name: 'month', intro: 'This month, from the first.' },
    since_all: { name: 'all', intro: 'Everything, however old.' },
  },
};

{
  const fm_sinceStart = setInterval(() => {
    if (typeof feature_Loop !== 'undefined') {
      clearInterval(fm_sinceStart);
      feature_Since.tell();
    }
  }, 100);
  if (typeof document !== 'undefined') {
    document.addEventListener('visibilitychange', () => {
      if (document.visibilityState === 'visible') feature_Since.tell();
    });
  }
  // the pills are not tool buttons, so /sub-tool-cards does not arm them; arm
  // them the way /tool-words arms the picker's, and answer with this node's
  // own words before anything else in the chain.
  if (typeof feature_LongPress !== 'undefined') {
    const fm_sinceContent = feature_LongPress.contentFor.bind(feature_LongPress);
    feature_LongPress.contentFor = async function (btn) {
      const w = feature_Since.WORDS[btn.getAttribute('data-ev') || ''];
      if (w) return { name: w.name, intro: w.intro };
      return fm_sinceContent(btn);
    };
    document.addEventListener('pointerdown', (e) => {
      const btn = e.target && e.target.closest ? e.target.closest('.since-pill[data-ev]') : null;
      if (!btn) return;
      feature_LongPress.disarm();
      feature_LongPress.fired = false;
      feature_LongPress.armed = btn;
      feature_LongPress.x = e.clientX;
      feature_LongPress.y = e.clientY;
      feature_LongPress.timer = setTimeout(() => feature_LongPress.show(btn), 500);
    });
    // a long press READS; the tap that follows it must not also switch.
    document.addEventListener('click', (e) => {
      if (feature_LongPress.fired && e.target && e.target.closest && e.target.closest('.since-pill[data-ev]')) {
        e.stopPropagation();
        e.preventDefault();
        feature_LongPress.fired = false;
      }
    }, true);
  }
}

File: features/miso/loop/cards/browse/map-only/since/since.rs
struct feature_Since;
impl feature_Since {
    // ---- the two vars ------------------------------------------------------
    // read through the live context rather than the bridged loop state, for
    // /browse's own reason: /payload republishes part-way down the update
    // chain and this node's links sit outside it, so `s.<key>` in a render
    // that follows this node's own write would be one turn stale.

    fn since_period_read() -> String {
        with_context(|c| c.since_period_get())
    }

    fn since_period_write(period: String) {
        edit_context(|c| {
            let _ = c.edit_op("miso/loop/cards/browse/map-only/since", "period",
                              serde_json::json!(period.clone()));
        });
    }

    fn since_marks_read() -> String {
        with_context(|c| c.since_day_starts_get())
    }

    fn since_marks_write(marks: String) {
        edit_context(|c| {
            let _ = c.edit_op("miso/loop/cards/browse/map-only/since",
                              "day_starts", serde_json::json!(marks.clone()));
        });
    }

    // ---- the pills ---------------------------------------------------------
    // /map-only emptied the picker's slot and left this seam for whatever takes
    // it. Four pills in /audience's grade-pill grammar: a small closed set,
    // round, dim, one lit in the accent that already means CHOSEN (/taste 3).

    fn browse_slot_html() -> String {
        let p = since_period_read();
        format!("<div class=\"since-pills\">{}{}{}{}</div>",
                since_pill("today".to_string(), p == "today"),
                since_pill("week".to_string(), p == "week"),
                since_pill("month".to_string(), p == "month"),
                since_pill("all".to_string(), p != "today" && p != "week" && p != "month"))
    }

    // the four data-ev strings are written out rather than formatted, because
    // /sub-tool-cards' and /tool-words' long press read them out of this source
    // to name the control a finger is held on, and skip any that carry a
    // format placeholder.
    fn since_pill(which: String, on: bool) -> String {
        let lit = if on { " since-on" } else { "" };
        if which == "today" {
            return format!("<div class=\"since-pill{}\" data-ev=\"since_today\" title=\"today\">today</div>", lit);
        }
        if which == "week" {
            return format!("<div class=\"since-pill{}\" data-ev=\"since_week\" title=\"week\">week</div>", lit);
        }
        if which == "month" {
            return format!("<div class=\"since-pill{}\" data-ev=\"since_month\" title=\"month\">month</div>", lit);
        }
        format!("<div class=\"since-pill{}\" data-ev=\"since_all\" title=\"all\">all</div>", lit)
    }

    // ---- the events --------------------------------------------------------
    // a pill writes the period and leaves `open` alone: the picker cleared it
    // because a mode you cannot see is a control doing nothing, but narrowing
    // the set behind a card you are reading is visible the moment you go back.
    //
    // SinceMarks is the page half's answer to "when was local midnight" — an
    // event, the way a finger sends one, not a write to a bridged key
    // (misses.md, navigation from the wrong side).

    fn update(state: String, event: String) -> String {
        let state = existing.update(state, event.clone());
        let e: serde_json::Value = serde_json::from_str(&event)
            .unwrap_or(serde_json::Value::Null);
        let kind = e["type"].as_str().unwrap_or("").to_string();
        if kind == "SinceMarks" {
            let marks = e["data"]["marks"].as_str().unwrap_or("").to_string();
            if !marks.is_empty() && marks != since_marks_read() {
                since_marks_write(marks);
            }
            return state;
        }
        if kind != "click" {
            return state;
        }
        let ev = e["ev"].as_str().unwrap_or("").to_string();
        if ev == "since_today" || ev == "since_week" || ev == "since_month"
            || ev == "since_all" {
            since_period_write(ev["since_".len()..].to_string());
        }
        state
    }

    // ---- the cut -----------------------------------------------------------
    // one epoch millisecond floor, and 0 means keep everything. 0 is the answer
    // for `all`, for a period whose mark is missing, and for marks that have
    // not arrived — a filter that has not been told the time shows you your
    // world rather than an empty map.

    fn since_cut() -> u64 {
        let p = since_period_read();
        let at = if p == "today" {
            0
        } else if p == "week" {
            1
        } else if p == "month" {
            2
        } else {
            return 0;
        };
        let marks = since_marks_read();
        let parts: Vec<String> = marks.split(',').map(|s| s.to_string()).collect();
        if parts.len() < 3 {
            return 0;
        }
        parts[at].parse::<u64>().unwrap_or(0)
    }

    // the card's own moment: a post's `when` if it has one, and otherwise the
    // moment the card was made. The field is read straight off the card rather
    // than through /post-time's function, so this keeps working with that node
    // unticked — no card then carries a `when` it wrote, and everything falls
    // back to `created`, which is what the ask says a person and a project
    // count by anyway.
    fn since_time_of(card: &serde_json::Value) -> u64 {
        let w = card["when"].as_u64().unwrap_or(0);
        if w > 0 {
            return w;
        }
        card["created"].as_u64().unwrap_or(0)
    }

    // the test, with the two exemptions the spec names: your own profile card
    // (a list of people that does not contain you is somebody else's list),
    // and the card that is open (narrow the band behind it, do not close it).
    fn since_keep(card: &serde_json::Value) -> bool {
        let cut = since_cut();
        if cut == 0 {
            return true;
        }
        if card["type"].as_str().unwrap_or("") == "profile"
            && card["from"].is_null() {
            return true;
        }
        // the open card, but only while a tool is actually drawing it: `open`
        // outlives the tool that set it (/new writes it from the launcher, and
        // nothing clears it until the next tool tap), and an exemption that
        // did not check would leak one card into every surface's set for as
        // long as it sat there. Caught on the rig, 2026-09-04.
        let id = card["id"].as_str().unwrap_or("").to_string();
        if !id.is_empty() && id == browse_open_read()
            && !open_tool_read().is_empty() {
            return true;
        }
        since_time_of(card) >= cut
    }

    // ---- the two chains that carry a browsed set ----------------------------
    // /browse's seam (which /people and /projects take) and /posts' own set.
    // Narrowing here is upstream of everything that draws: the map's pins, the
    // band's data-ids, /on-people-map's data-post-ids and /flick's walk all
    // read what these two return, so they agree without being told and /reel's
    // own contract — the band lists the map's set — is untouched.

    fn browse_cards(state: String) -> String {
        let list: serde_json::Value =
            serde_json::from_str(&existing.browse_cards(state))
                .unwrap_or(serde_json::json!([]));
        let empty: Vec<serde_json::Value> = Vec::new();
        let mut out: Vec<serde_json::Value> = Vec::new();
        for c in list.as_array().unwrap_or(&empty) {
            if since_keep(c) {
                out.push(c.clone());
            }
        }
        serde_json::Value::Array(out).to_string()
    }

    fn posts_set() -> Vec<serde_json::Value> {
        let mut out: Vec<serde_json::Value> = Vec::new();
        for c in existing.posts_set().iter() {
            if since_keep(c) {
                out.push(c.clone());
            }
        }
        out
    }
}
